src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -265,67 +266,81 @@ where
         let session = self
             .session_mgr
             .connect(&db_name, &user_name)
-            .map_err(|e| StartupError::ConnectError(e))?;
+            .map_err(PsqlError::StartupError)?;
         match session.user_authenticator() {
             UserAuthenticator::None => {
-                self.write_message_no_flush(&BeMessage::AuthenticationOk)?;
-                self.write_parameter_status_msg_no_flush()?;
-                self.write_message_no_flush(&BeMessage::ReadyForQuery)?;
+                self.write_message_no_flush(&BeMessage::AuthenticationOk)
+                    .map_err(|err| PsqlError::StartupError(Box::new(err)))?;
+                self.write_parameter_status_msg_no_flush()
+                    .map_err(|err| PsqlError::StartupError(Box::new(err)))?;
+                self.write_message_no_flush(&BeMessage::ReadyForQuery)
+                    .map_err(|err| PsqlError::StartupError(Box::new(err)))?;
             }
             UserAuthenticator::ClearText(_) => {
-                self.write_message_no_flush(&BeMessage::AuthenticationCleartextPassword)?;
+                self.write_message_no_flush(&BeMessage::AuthenticationCleartextPassword)
+                    .map_err(|err| PsqlError::StartupError(Box::new(err)))?;
             }
             UserAuthenticator::MD5WithSalt { salt, .. } => {
-                self.write_message_no_flush(&BeMessage::AuthenticationMD5Password(salt))?;
+                self.write_message_no_flush(&BeMessage::AuthenticationMD5Password(salt))
+                    .map_err(|err| PsqlError::StartupError(Box::new(err)))?;
             }
         }
         self.session = Some(session);
         Ok(())
     }
 
-    fn process_password_msg(&mut self, msg: FePasswordMessage) -> Result<(), PasswordError> {
+    fn process_password_msg(&mut self, msg: FePasswordMessage) -> PsqlResult<()> {
         let authenticator = self.session.as_ref().unwrap().user_authenticator();
         if !authenticator.authenticate(&msg.password) {
-            return Err(PasswordError::InvalidPassword);
+            return Err(PsqlError::PasswordError(IoError::new(
+                ErrorKind::InvalidInput,
+                "Invalid password",
+            )));
         }
-        self.write_message_no_flush(&BeMessage::AuthenticationOk)?;
-        self.write_parameter_status_msg_no_flush()?;
-        self.write_message_no_flush(&BeMessage::ReadyForQuery)?;
+        self.write_message_no_flush(&BeMessage::AuthenticationOk)
+            .map_err(PsqlError::PasswordError)?;
```

Comment:
For write_message or write_message_no_flush, maybe we should not `.map_err` distinguish different type of error and all treat them like PsqlError::IoError?

Reply:
write_message_no_flush in process_password and process_startup stiil need to map as PasswordError and StartupError. Because in process_password/process_startup, if write_message_no_flush return a error, you need to close a connection rather than just report it. So we can't treat it as a PsqlError::IoError, PsqlError::IoError will just report but not close connection.

Reply:
Actually I think every Io::Error should close connection. Let me take a look

Reply:
In past version, I think most of Io::Error don't close connection.